[runtime env] `conda` does not work with Python 3.10

### What happened + What you expected to happen

The `conda` field of `runtime_env` internally inserts the appropriate `ray` wheel into the conda environment.  The code to do this hasn't been updated to support Python 3.10.  Python 3.10 is currently "experimental" for Ray but we still need to fix this eventually.

```
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/utils.py", line 1264, in get_wheel_filename
    assert py_version in ["36", "37", "38", "39"], py_version
```

We should update this, and ideally add this step to the release process for new Ray Python versions (or otherwise automate this somehow)

### Versions / Dependencies

Python 3.10

### Reproduction script

```
from ray.runtime_env import RuntimeEnv
runtime_env = RuntimeEnv(
    conda={
        "dependencies":
        ["toolz", "dill", "pip", {
            "pip": ["pendulum", "ray[serve]"]
        }]
    },
    env_vars={"TF_WARNINGS": "none"}
)
ray.init(runtime_env=runtime_env)
@ray.remote
def f():
    pass

ray.get(f.remote())
```

Gives the traceback clearly indicating the error
```
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/util/client/server/proxier.py", line 703, in Datapath
    if not self.proxy_manager.start_specific_server(
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/util/client/server/proxier.py", line 304, in start_specific_server
    serialized_runtime_env_context = self._create_runtime_env(
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/util/client/server/proxier.py", line 253, in _create_runtime_env
    raise RuntimeError(
RuntimeError: Failed to create runtime_env for Ray client server, it is caused by:
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 342, in _create_runtime_env_with_retry
    runtime_env_context = await asyncio.wait_for(
  File "/home/ray/anaconda3/lib/python3.10/asyncio/tasks.py", line 445, in wait_for
    return fut.result()
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 297, in _setup_runtime_env
    await create_for_plugin_if_needed(
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/runtime_env/plugin.py", line 252, in create_for_plugin_if_needed
    size_bytes = await plugin.create(uri, runtime_env, context, logger=logger)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/runtime_env/conda.py", line 362, in create
    return await loop.run_in_executor(None, _create)
  File "/home/ray/anaconda3/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/runtime_env/conda.py", line 340, in _create
    conda_dict = _get_conda_dict_with_ray_inserted(runtime_env, logger=logger)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/runtime_env/conda.py", line 244, in _get_conda_dict_with_ray_inserted
    ray_pip = current_ray_pip_specifier(logger=logger)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/runtime_env/conda.py", line 156, in current_ray_pip_specifier
    return get_release_wheel_url()
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/utils.py", line 1308, in get_release_wheel_url
    filename = get_wheel_filename(
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/utils.py", line 1264, in get_wheel_filename
    assert py_version in ["36", "37", "38", "39"], py_version
AssertionError: 310
```

### Issue Severity

None

--- a/python/ray/_private/ray_constants.py
+++ b/python/ray/_private/ray_constants.py
@@ -364,4 +364,7 @@ def gcs_actor_scheduling_enabled():
 
 DEFAULT_RESOURCES = {"CPU", "GPU", "memory", "object_store_memory"}
 
-SUPPORTED_PYTHON_VERSIONS = [(3, 6), (3, 7), (3, 8), (3, 9), (3, 10)]
+# Supported Python versions for runtime env's "conda" field. Ray downloads
+# Ray wheels into the conda environment, so the Ray wheels for these Python
+# versions must be available online.
+RUNTIME_ENV_CONDA_PY_VERSIONS = [(3, 6), (3, 7), (3, 8), (3, 9), (3, 10)]

--- a/python/ray/_private/ray_constants.py
+++ b/python/ray/_private/ray_constants.py
@@ -363,3 +363,5 @@ def gcs_actor_scheduling_enabled():
 
 
 DEFAULT_RESOURCES = {"CPU", "GPU", "memory", "object_store_memory"}
+
+SUPPORTED_PYTHON_VERSIONS = [(3, 6), (3, 7), (3, 8), (3, 9), (3, 10)]

--- a/python/ray/_private/utils.py
+++ b/python/ray/_private/utils.py
@@ -1262,7 +1262,7 @@ def get_wheel_filename(
         The wheel file name.  Examples:
             ray-3.0.0.dev0-cp38-cp38-manylinux2014_x86_64.whl
     """
-    assert py_version in ray_constants.SUPPORTED_PYTHON_VERSIONS, py_version
+    assert py_version in ray_constants.RUNTIME_ENV_CONDA_PY_VERSIONS, py_version
 
     py_version_str = "".join(map(str, py_version))
     if py_version_str in ["36", "37"]:

--- a/python/ray/_private/utils.py
+++ b/python/ray/_private/utils.py
@@ -1247,7 +1247,7 @@ def import_attr(full_path: str):
 def get_wheel_filename(
     sys_platform: str = sys.platform,
     ray_version: str = ray.__version__,
-    py_version: Tuple[int, int] = sys.version_info[:2],
+    py_version: Tuple[int, int] = (sys.version_info.major, sys.version_info.minor),
 ) -> str:
     """Returns the filename used for the nightly Ray wheel.
 

--- a/python/ray/_private/utils.py
+++ b/python/ray/_private/utils.py
@@ -1262,7 +1262,7 @@ def get_wheel_filename(
         The wheel file name.  Examples:
             ray-3.0.0.dev0-cp38-cp38-manylinux2014_x86_64.whl
     """
-    assert py_version in ray_constants.SUPPORTED_PY_VERSIONS, py_version
+    assert py_version in ray_constants.SUPPORTED_PYTHON_VERSIONS, py_version
 
     py_version_str = "".join(map(str, py_version))
     if py_version_str in ["36", "37"]:

--- a/python/ray/_private/utils.py
+++ b/python/ray/_private/utils.py
@@ -1266,9 +1266,9 @@ def get_wheel_filename(
     assert py_version in ["36", "37", "38", "39", "310"], py_version
 
     os_strings = {
-        "darwin": "macosx_10_15_x86_64"
-        if py_version in ["38", "39", "310"]
-        else "macosx_10_15_intel",
+        "darwin": "macosx_10_15_intel"
+        if py_version in ["36", "37"]
+        else "macosx_10_15_x86_64",
         "linux": "manylinux2014_x86_64",
         "win32": "win_amd64",
     }

--- a/python/ray/_private/utils.py
+++ b/python/ray/_private/utils.py
@@ -1247,7 +1247,7 @@ def import_attr(full_path: str):
 def get_wheel_filename(
     sys_platform: str = sys.platform,
     ray_version: str = ray.__version__,
-    py_version: str = f"{sys.version_info.major}{sys.version_info.minor}",
+    py_version: Tuple[int, int] = sys.version_info[:2],
 ) -> str:
     """Returns the filename used for the nightly Ray wheel.
 
@@ -1256,19 +1256,23 @@ def get_wheel_filename(
             "darwin", "linux", "win32"
         ray_version: The Ray version as returned by ray.__version__ or
             `ray --version`.  Examples: "3.0.0.dev0"
-        py_version (str):
-            The major and minor Python versions concatenated.  Examples: "36",
-            "37", "38", "39"
+        py_version: The Python version as returned by sys.version_info. A
+            tuple of (major, minor). Examples: (3, 8)
     Returns:
         The wheel file name.  Examples:
             ray-3.0.0.dev0-cp38-cp38-manylinux2014_x86_64.whl
     """
-    assert py_version in ["36", "37", "38", "39", "310"], py_version
+    assert py_version in ray_constants.SUPPORTED_PY_VERSIONS, py_version
 
+    py_version_str = "".join(map(str, py_version))
+    if py_version_str in ["36", "37"]:
+        darwin_os_string = "macosx_10_15_intel"
+    elif py_version_str in ["38", "39"]:
+        darwin_os_string = "macosx_10_15_x86_64"
+    else:
+        darwin_os_string = "macosx_10_15_universal2"
     os_strings = {
-        "darwin": "macosx_10_15_intel"
-        if py_version in ["36", "37"]
-        else "macosx_10_15_x86_64",
+        "darwin": darwin_os_string,
         "linux": "manylinux2014_x86_64",
         "win32": "win_amd64",
     }
@@ -1276,8 +1280,8 @@ def get_wheel_filename(
     assert sys_platform in os_strings, sys_platform
 
     wheel_filename = (
-        f"ray-{ray_version}-cp{py_version}-"
-        f"cp{py_version}{'m' if py_version in ['36', '37'] else ''}"
+        f"ray-{ray_version}-cp{py_version_str}-"
+        f"cp{py_version_str}{'m' if py_version_str in ['36', '37'] else ''}"
         f"-{os_strings[sys_platform]}.whl"
     )
 
@@ -1288,7 +1292,7 @@ def get_master_wheel_url(
     ray_commit: str = ray.__commit__,
     sys_platform: str = sys.platform,
     ray_version: str = ray.__version__,
-    py_version: str = f"{sys.version_info.major}{sys.version_info.minor}",
+    py_version: Tuple[int, int] = sys.version_info[:2],
 ) -> str:
     """Return the URL for the wheel from a specific commit."""
     filename = get_wheel_filename(
@@ -1304,7 +1308,7 @@ def get_release_wheel_url(
     ray_commit: str = ray.__commit__,
     sys_platform: str = sys.platform,
     ray_version: str = ray.__version__,
-    py_version: str = f"{sys.version_info.major}{sys.version_info.minor}",
+    py_version: Tuple[int, int] = sys.version_info[:2],
 ) -> str:
     """Return the URL for the wheel for a specific release."""
     filename = get_wheel_filename(

--- a/python/ray/_private/utils.py
+++ b/python/ray/_private/utils.py
@@ -1263,11 +1263,11 @@ def get_wheel_filename(
         The wheel file name.  Examples:
             ray-3.0.0.dev0-cp38-cp38-manylinux2014_x86_64.whl
     """
-    assert py_version in ["36", "37", "38", "39"], py_version
+    assert py_version in ["36", "37", "38", "39", "310"], py_version
 
     os_strings = {
         "darwin": "macosx_10_15_x86_64"
-        if py_version in ["38", "39"]
+        if py_version in ["38", "39", "310"]
         else "macosx_10_15_intel",
         "linux": "manylinux2014_x86_64",
         "win32": "win_amd64",

--- a/python/ray/tests/test_runtime_env.py
+++ b/python/ray/tests/test_runtime_env.py
@@ -45,7 +45,7 @@ def test_get_wheel_filename():
     # NOTE: These should not be changed for releases.
     ray_version = "3.0.0.dev0"
     for sys_platform in ["darwin", "linux", "win32"]:
-        for py_version in [(3, 6), (3, 7), (3, 8), (3, 9), (3, 10)]:
+        for py_version in ray_constants.RUNTIME_ENV_CONDA_PY_VERSIONS:
             if sys_platform == "win32" and py_version == (3, 6):
                 # Windows wheels are not built for py3.6 anymore
                 continue
@@ -64,7 +64,7 @@ def test_get_master_wheel_url():
     # `s3://ray-wheels/master/<test_commit>/`.
     test_commit = "6fd684bbdb186a73732f6113a83a12b63200f170"
     for sys_platform in ["darwin", "linux", "win32"]:
-        for py_version in [(3, 6), (3, 7), (3, 8), (3, 9), (3, 10)]:
+        for py_version in ray_constants.RUNTIME_ENV_CONDA_PY_VERSIONS:
             if sys_platform == "win32" and py_version == (3, 6):
                 # Windows wheels are not built for py3.6 anymore
                 continue
@@ -81,7 +81,7 @@ def test_get_release_wheel_url():
     # `s3://ray-wheels/releases/2.2.0/<commit>/`.
     test_commits = {"2.2.0": "b6af0887ee5f2e460202133791ad941a41f15beb"}
     for sys_platform in ["darwin", "linux", "win32"]:
-        for py_version in [(3, 6), (3, 7), (3, 8), (3, 9), (3, 10)]:
+        for py_version in ray_constants.RUNTIME_ENV_CONDA_PY_VERSIONS:
             for version, commit in test_commits.items():
                 if sys_platform == "win32" and py_version == (3, 6):
                     # Windows wheels are not built for py3.6 anymore
@@ -99,14 +99,14 @@ def test_current_py_version_supported():
     the `conda` environment installation will fail.
 
     When a new python version is added to the Ray wheels, please update
-    `ray_constants.SUPPORTED_PYTHON_VERSIONS`.  In a subsequent commit,
+    `ray_constants.RUNTIME_ENV_CONDA_PY_VERSIONS`.  In a subsequent commit,
     once wheels have been built for the new python version, please update
     the tests test_get_wheel_filename, test_get_master_wheel_url, and
     (after the first Ray release with the new python version)
     test_get_release_wheel_url.
     """
     py_version = sys.version_info[:2]
-    assert py_version in ray_constants.SUPPORTED_PYTHON_VERSIONS
+    assert py_version in ray_constants.RUNTIME_ENV_CONDA_PY_VERSIONS
 
 
 def test_compatible_with_dataclasses():

--- a/python/ray/tests/test_runtime_env.py
+++ b/python/ray/tests/test_runtime_env.py
@@ -37,13 +37,16 @@ from ray._private.utils import (
 from ray.exceptions import RuntimeEnvSetupError
 from ray.runtime_env import RuntimeEnv
 
+import ray._private.ray_constants as ray_constants
+
 
 def test_get_wheel_filename():
+    """Test the code that generates the filenames of the `latest` wheels."""
     # NOTE: These should not be changed for releases.
     ray_version = "3.0.0.dev0"
     for sys_platform in ["darwin", "linux", "win32"]:
-        for py_version in ["36", "37", "38", "39", "310"]:
-            if sys_platform == "win32" and py_version == "36":
+        for py_version in [(3, 6), (3, 7), (3, 8), (3, 9), (3, 10)]:
+            if sys_platform == "win32" and py_version == (3, 6):
                 # Windows wheels are not built for py3.6 anymore
                 continue
             filename = get_wheel_filename(sys_platform, ray_version, py_version)
@@ -53,15 +56,16 @@ def test_get_wheel_filename():
 
 
 def test_get_master_wheel_url():
+    """Test the code that generates the filenames of `master` commit wheels."""
     # NOTE: These should not be changed for releases.
     ray_version = "3.0.0.dev0"
-    # This should be a commit for which wheels have been built for
+    # This should be a commit for which wheels have already been built for
     # all platforms and python versions at
     # `s3://ray-wheels/master/<test_commit>/`.
     test_commit = "6fd684bbdb186a73732f6113a83a12b63200f170"
     for sys_platform in ["darwin", "linux", "win32"]:
-        for py_version in ["36", "37", "38", "39", "310"]:
-            if sys_platform == "win32" and py_version == "36":
+        for py_version in [(3, 6), (3, 7), (3, 8), (3, 9), (3, 10)]:
+            if sys_platform == "win32" and py_version == (3, 6):
                 # Windows wheels are not built for py3.6 anymore
                 continue
             url = get_master_wheel_url(
@@ -71,16 +75,38 @@ def test_get_master_wheel_url():
 
 
 def test_get_release_wheel_url():
-    # This should be a commit for which wheels have been built for
+    """Test the code that generates the filenames of the `release` branch wheels."""
+    # This should be a commit for which wheels have already been built for
     # all platforms and python versions at
     # `s3://ray-wheels/releases/2.2.0/<commit>/`.
-    # NOTE: These should not be changed for releases.
-    test_commits = {"2.2.0": "03de294293ab598da6e183f182b49b7003cceba8"}
+    test_commits = {"2.2.0": "b6af0887ee5f2e460202133791ad941a41f15beb"}
     for sys_platform in ["darwin", "linux", "win32"]:
-        for py_version in ["36", "37", "38", "39", "310"]:
+        for py_version in [(3, 6), (3, 7), (3, 8), (3, 9), (3, 10)]:
             for version, commit in test_commits.items():
+                if sys_platform == "win32" and py_version == (3, 6):
+                    # Windows wheels are not built for py3.6 anymore
+                    continue
                 url = get_release_wheel_url(commit, sys_platform, version, py_version)
                 assert requests.head(url).status_code == 200, url
+
+
+def test_current_py_version_supported():
+    """Test that the running python version is supported.
+
+    This is run as a check in the Ray `runtime_env` `conda` code
+    before downloading the Ray wheel into the conda environment.
+    If Ray wheels are not available for this python version, then
+    the `conda` environment installation will fail.
+
+    When a new python version is added to the Ray wheels, please update
+    `ray_constants.SUPPORTED_PYTHON_VERSIONS`.  In a subsequent commit,
+    once wheels have been built for the new python version, please update
+    the tests test_get_wheel_filename, test_get_master_wheel_url, and
+    (after the first Ray release with the new python version)
+    test_get_release_wheel_url.
+    """
+    py_version = sys.version_info[:2]
+    assert py_version in ray_constants.SUPPORTED_PYTHON_VERSIONS
 
 
 def test_compatible_with_dataclasses():

--- a/python/ray/tests/test_runtime_env.py
+++ b/python/ray/tests/test_runtime_env.py
@@ -42,7 +42,7 @@ def test_get_wheel_filename():
     # NOTE: These should not be changed for releases.
     ray_version = "3.0.0.dev0"
     for sys_platform in ["darwin", "linux", "win32"]:
-        for py_version in ["36", "37", "38", "39"]:
+        for py_version in ["36", "37", "38", "39", "310"]:
             if sys_platform == "win32" and py_version == "36":
                 # Windows wheels are not built for py3.6 anymore
                 continue
@@ -55,9 +55,12 @@ def test_get_wheel_filename():
 def test_get_master_wheel_url():
     # NOTE: These should not be changed for releases.
     ray_version = "3.0.0.dev0"
-    test_commit = "c3ac6fcf3fcc8cfe6930c9a820add0e187bff579"
+    # This should be a commit for which wheels have been built for
+    # all platforms and python versions at
+    # `s3://ray-wheels/master/<test_commit>/`.
+    test_commit = "6fd684bbdb186a73732f6113a83a12b63200f170"
     for sys_platform in ["darwin", "linux", "win32"]:
-        for py_version in ["36", "37", "38", "39"]:
+        for py_version in ["36", "37", "38", "39", "310"]:
             if sys_platform == "win32" and py_version == "36":
                 # Windows wheels are not built for py3.6 anymore
                 continue
@@ -68,9 +71,13 @@ def test_get_master_wheel_url():
 
 
 def test_get_release_wheel_url():
-    test_commits = {"1.6.0": "5052fe67d99f1d4bfc81b2a8694dbf2aa807bbdc"}
+    # This should be a commit for which wheels have been built for
+    # all platforms and python versions at
+    # `s3://ray-wheels/releases/2.2.0/<commit>/`.
+    # NOTE: These should not be changed for releases.
+    test_commits = {"2.2.0": "03de294293ab598da6e183f182b49b7003cceba8"}
     for sys_platform in ["darwin", "linux", "win32"]:
-        for py_version in ["36", "37", "38", "39"]:
+        for py_version in ["36", "37", "38", "39", "310"]:
             for version, commit in test_commits.items():
                 url = get_release_wheel_url(commit, sys_platform, version, py_version)
                 assert requests.head(url).status_code == 200, url

--- a/release/runtime_env_tests/workloads/wheel_urls.py
+++ b/release/runtime_env_tests/workloads/wheel_urls.py
@@ -40,7 +40,7 @@ if __name__ == "__main__":
 
     retry = set()
     for sys_platform in ["darwin", "linux", "win32"]:
-        for py_version in ["36", "37", "38", "39"]:
+        for py_version in ["36", "37", "38", "39", "310"]:
             if "dev" in ray.__version__:
                 url = get_master_wheel_url(
                     ray_commit=ray.__commit__,

